presto-kudu/src/main/java/io/prestosql/plugin/kudu/KuduSplitManager.java

```diff
@@ -59,4 +74,72 @@ public ConnectorSplitSource getSplits(
                 throw new IllegalArgumentException("Unknown splitSchedulingStrategy: " + splitSchedulingStrategy);
         }
     }
+
+    private class KuduDynamicFilteringSplitSource
+            implements ConnectorSplitSource
+    {
+        private final CompletableFuture<ConnectorSplitSource> future;
+
+        private KuduDynamicFilteringSplitSource(
+                ConnectorTableHandle table,
+                SplitSchedulingStrategy splitSchedulingStrategy,
+                DynamicFilter dynamicFilter,
+                long blockingTimeoutMillis)
+        {
+            this.future = whenCompleted(dynamicFilter)
```

Comment:
DF could be narrowed down multiple times, please use approach as in `io.prestosql.plugin.hive.BackgroundHiveSplitLoader.HiveSplitLoaderTask#process`

Reply:
whenCompleted() is a recursive method which wait for narrowing multiple times until not awaitable.